enterprise/backend/src/metabase_enterprise/sandbox/api/user.clj

```diff
@@ -28,17 +28,19 @@
   (api/check-404 (t2/select-one :model/User :id id))
   (pos? (t2/update! :model/User id {:login_attributes login_attributes})))
 
+(def ^:private max-login-attributes 5000)
+
 (api.macros/defendpoint :get "/attributes"
   "Fetch a list of possible keys for User `login_attributes`. This just looks at keys that have already been set for
   existing Users and returns those. "
   []
-  (->>
-   ;; look at the `login_attributes` for the first 1000 users that have them set. Then make a set of the keys
-   (for [login-attributes (t2/select-fn-set :login_attributes :model/User :login_attributes [:not= nil] {:limit 1000})
-         :when (seq login-attributes)]
-     (set (keys login-attributes)))
-   ;; combine all the sets of attribute keys into a single set
-   (reduce set/union #{})))
+  (->> (t2/select-fn-set :login_attributes :model/User {:where [:and
+                                                                [:not= :login_attributes nil]
+                                                                [:not= :login_attributes "{}"]]})
+       (mapcat keys)
```

Comment:
I think mapcat is not lazy, so am wondering if this chain cloud cause us to load a bunch db rows into memory if this table is particularly large? I ran into that in the analytics module.

If we do (map keys) (flatten) would that make sure we only realize enough rows from the db to satisfy the take?

Reply:
Actually, `mapcat` does seem to be lazy:

```
(type (mapcat identity (range 1000)))
;; => clojure.lang.LazySeq
```

So I decided to just make this lazier across the board using `t2/select-fn-reducible`

Reply:
neat, had it in my head it was the same as `(apply concat (map fn` so was worried it was going to realize everything